gammapy/spectrum/energy_group.py

```diff
@@ -0,0 +1,534 @@
+UNDERFLOW_BIN_INDEX = -1
+OVERFLOW_BIN_INDEX = -2
+
+
+class SpectrumEnergyGroupMaker(object):
+    """Energy bin groups for spectral analysis.
+
+    This class contains both methods that run algorithms
+    that compute groupings as well as the results as data members
+    and methods to debug and assess the results.
+
+    The input ``obs`` should be used read-only.
+    All computations modify ``table`` or other properties.
+
+    TODO: should this class have a `SpectrumEnergyGroups` data member
+    instead of or in addition to `table`?
+
+    See :ref:`spectrum_energy_group` for examples.
+
+    Attributes
+    ----------
+    obs : `~gammapy.spectrum.SpectrumObservation`
+        Spectrum observation data
+    table : `~astropy.table.QTable`
+        Table with some per-energy bin stats info.
+    groups : `~gammapy.spectrum.SpectrumEnergyGroups`
+        List of energy groups.
+    """
+
+    def __init__(self, obs=None):
+        self.obs = obs
+
+        # Start with a table with the energy binning and basic stats
+        self.table = self.obs.stats_table()
+        # Start out with one bin = one group
+        self.table['bin_idx'] = np.arange(len(self.table))
+        self.table['energy_group_idx'] = self.table['bin_idx']
+
+        # The algorithms use this `groups` object
+        # Unfortunately one has to be careful to copy the results
+        # back into the `table` object to keep the two in sync
+        self.groups = SpectrumEnergyGroups.from_total_table(self.table)
+
+    def __str__(self):
+        """Print a little report"""
+        ss = 'SpectrumEnergyGroupMaker:\n'
+
+        ss += '\nSpectrum table:\n'
+        ss += 'Number of bins: {}\n'.format(len(self.table))
+        ss += 'Bin range: {}\n'.format((0, len(self.table) - 1))
+        ss += 'Energy range: {}\n'.format(self.table_energy_range)
+
+        ss += str(self.groups)
+
+        return ss
+
+    # Properties concerning the total, un-grouped spectrum
+
+    @property
+    def table_energy_range(self):
+        """Total spectrum energy range (no grouping or range applied)"""
+        emin = self.table['energy_min'][0]
+        emax = self.table['energy_max'][-1]
+        return EnergyRange(emin, emax)
+
+    # Properties for the grouped spectrum
+
+    @property
+    def n_groups(self):
+        """Number of groups."""
+        return len(self.groups)
+
+    # Methods to compute total ranges
+
+    def compute_range_safe(self):
+        """Apply safe energy range to ``groups``.
+        """
+        emin = self.obs.lo_threshold
+        emax = self.obs.hi_threshold
+        self.set_energy_range(emin=emin, emax=emax)
+
+    def set_energy_range(self, emin=None, emax=None):
+        """Apply energy range to ``groups``.
+        """
+        if emin:
+            self.groups.apply_energy_min(emin)
+        if emax:
+            self.groups.apply_energy_max(emax)
+
+    # Methods to compute groupings
+
+    def compute_groups_fixed(self, ebounds):
```

Comment:
The stacked spectrum object does not have data outside the safe energy range. What does this method do if ebounds[0] < obs.lo_threshold?